check_diff/src/lib.rs

```diff
@@ -56,3 +136,139 @@ pub fn change_directory_to_path(dest: &Path) -> io::Result<()> {
+        return Err(CheckDiffError::FailedCargoVersion(
+            "Failed to obtain cargo version".to_string(),
+        ));
+    };
+
+    let Ok(cargo_version) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting cargo version to string".to_string(),
+        ));
+    };
+
+    return Ok(cargo_version);
+}
+
+pub fn copy_src_to_dst(src: &str, dst: &str) -> Result<(), CheckDiffError> {
+    let Ok(_) = Command::new("cp").args([src, dst]).output() else {
+        return Err(CheckDiffError::FailedCopy(
+            "Error copying rustfmt release to destination".to_string(),
+        ));
+    };
+    return Ok(());
+}
+
+pub fn build_rsfmt_from_src(ld_lib_path: &String) -> Result<(), CheckDiffError> {
+    let Ok(_) = Command::new("cargo")
+        .env("LD_LIB_PATH", ld_lib_path)
+        .args(["build", "-q", "--release", "--bin", "rustfmt"])
+        .output()
+    else {
+        return Err(CheckDiffError::FailedSourceBuild(
+            "Error building rustfmt from source".to_string(),
+        ));
+    };
+    return Ok(());
+}
+
+pub fn get_binary_version(binary: String) -> Result<String, CheckDiffError> {
+    let Ok(command) = Command::new(binary.as_str()).args(["--version"]).output() else {
+        return Err(CheckDiffError::FailedVersioning(format!(
+            "Failed to get version for {}",
+            binary
+        )));
+    };
+
+    let Ok(binary_version) = String::from_utf8(command.stdout) else {
+        return Err(CheckDiffError::FailedUtf8(
+            "Error converting binary version to string".to_string(),
+        ));
+    };
+    return Ok(binary_version);
+}
+
+// Compiles and produces two rustfmt binaries.
+// One for the current master, and another for the feature branch
+// Parameters:
+// dest: Directory where rustfmt will be cloned
+pub fn compile_rustfmt(
+    dest: &Path,
+    remote_repo_url: String,
+    feature_branch: String,
+    commit_hash: Option<String>,
+) -> Result<Command, CheckDiffError> {
+    const RUSTFMT_REPO: &str = "https://github.com/rust-lang/rustfmt.git";
+
+    let _clone_git_result = clone_git_repo(RUSTFMT_REPO, dest)?;
+    let _git_remote_add_result = git_remote_add(remote_repo_url.as_str())?;
+    let _git_fetch_result = git_fetch(feature_branch.as_str())?;
+
+    let cargo_version = get_cargo_version()?;
+    info!("Compiling with {}", cargo_version);
+
+    //Because we're building standalone binaries we need to set `LD_LIBRARY_PATH` so each
+    // binary can find it's runtime dependencies.
+    // See https://github.com/rust-lang/rustfmt/issues/5675
+    // This will prepend the `LD_LIBRARY_PATH` for the master rustfmt binary
+    let ld_lib_path = get_ld_lib_path()?;
+
+    info!("Building rustfmt from source");
+    let _build_from_src = build_rsfmt_from_src(&ld_lib_path)?;
+
+    let rustfmt_binary = format!("{}/rustfmt", dest.display());
```

Comment:
Nicer to user [`Path::join`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.join) instead of [`Path::display`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.display).